ibis/mapd/client.py

```diff
@@ -0,0 +1,326 @@
+
+import regex as re
+import pandas as pd
+import pymapd
+
+import ibis.common as com
+import ibis.expr.types as ir
+import ibis.expr.datatypes as dt
+import ibis.expr.schema as sch
+
+fully_qualified_re = re.compile(r"(.*)\.(?:`(.*)`|(.*))")
+
+
+class MapDDataType(object):
+
+    __slots__ = 'typename', 'nullable'
+
+    # using impala.client._HS2_TTypeId_to_dtype as reference
+    # https://www.mapd.com/docs/latest/mapd-core-guide/fixed-encoding/
+    dtypes = {
+        'BIGINT': dt.int64,
+        'BOOLEAN': dt.Boolean,
+        'CHAR': dt.string,
+        'DATE': dt.date,
+        'DECIMAL': dt.float64,
+        'DOUBLE': dt.float64,
+        'INT': dt.int32,
+        'INTEGER': dt.int32,
+        'FLOAT': dt.float32,
+        'NULL': dt.Null,
+        'NUMERIC': dt.float64,
+        'REAL': dt.float32,
+        'SMALLINT': dt.int16,
+        'STR': dt.string,
+        'TEXT': dt.string,
+        'TIME': dt.time,
+        'TIMESTAMP': dt.timestamp,
+        'VARCHAR': dt.string,
+    }
+
+    ibis_dtypes = {
+        v: k for k, v in dtypes.items()
+    }
+
+    def __init__(self, typename, nullable=False):
+        if typename not in self.dtypes:
+            raise com.UnsupportedBackendType(typename)
+        self.typename = typename
+        self.nullable = nullable
+
+    def __str__(self):
+        if self.nullable:
+            return 'Nullable({})'.format(self.typename)
+        else:
+            return self.typename
+
+    def __repr__(self):
+        return '<MapD {}>'.format(str(self))
+
+    @classmethod
+    def parse(cls, spec):
+        if spec.startswith('Nullable'):
+            return cls(spec[9:-1], nullable=True)
+        else:
+            return cls(spec)
+
+    def to_ibis(self):
+        return self.dtypes[self.typename](nullable=self.nullable)
+
+    @classmethod
+    def from_ibis(cls, dtype, nullable=None):
+        dtype_ = type(dtype)
+        if dtype_ in cls.ibis_dtypes:
+            typename = cls.ibis_dtypes[dtype_]
+        elif dtype in cls.ibis_dtypes:
+            typename = cls.ibis_dtypes[dtype]
+        else:
+            raise NotImplemented('{0} not Implemented'.format(dtype))
+
+        if nullable is None:
+            nullable = dtype.nullable
+        return cls(typename, nullable=nullable)
+
+
+class MapDCursor(object):
+    """Cursor to allow the MapD client to reuse machinery in ibis/client.py
+    """
+
+    def __init__(self, cursor):
+        self.cursor = cursor
+
+    def to_df(self):
+        if isinstance(self.cursor, Cursor):
+            col_names = [c.name for c in self.cursor.description]
+            result = pd.DataFrame(self.cursor.fetchall(), columns=col_names)
+        elif self.cursor is None:
+            result = pd.DataFrame([])
+        elif isinstance(self.cursor, pd.DataFrame):
+            result = self.cursor
+        elif GPUDataFrame is not None and isinstance(
+                self.cursor, GPUDataFrame
+        ):
+            result = self.cursor
```

Comment:
Since you're executing the same code (assigning `self.cursor` to `result`) in the case that the cursor is a pandas DataFrame or a GPUDataFrame, can you remove the last two `elif`s? Is there a case where `self.cursor is not None` and it's not either a pandas DataFrame or a GPUDataFrame?

Reply:
you're right! thanks!